src/synthcity/plugins/core/models/tabular_ddpm/gaussian_multinomial_diffsuion.py

```diff
@@ -0,0 +1,1023 @@
+            raise NotImplementedError
+
+        model_mean, _, _ = self.gaussian_q_posterior_mean_variance(
+            x_start=pred_xstart, x_t=x, t=t
+        )
+
+        if not (
+            model_mean.shape == model_log_variance.shape == pred_xstart.shape == x.shape
+        ):
+            raise AssertionError(
+                f"{model_mean.shape}, {model_log_variance.shape}, {pred_xstart.shape}, {x.shape}"
+            )
+
+        return {
+            "mean": model_mean,
+            "variance": model_variance,
+            "log_variance": model_log_variance,
+            "pred_xstart": pred_xstart,
+        }
+
+    def _vb_terms_bpd(
+        self,
+        model_output: Tensor,
+        x_start: Tensor,
+        x_t: Tensor,
+        t: Tensor,
+        clip_denoised: bool = False,
+        model_kwargs: Optional[dict] = None,
+    ) -> dict:
+        (
+            true_mean,
+            _,
+            true_log_variance_clipped,
+        ) = self.gaussian_q_posterior_mean_variance(x_start=x_start, x_t=x_t, t=t)
+        out = self.gaussian_p_mean_variance(
+            model_output, x_t, t, clip_denoised=clip_denoised, model_kwargs=model_kwargs
+        )
+        kl = normal_kl(
+            true_mean, true_log_variance_clipped, out["mean"], out["log_variance"]
+        )
+        kl = mean_flat(kl) / np.log(2.0)
+
+        decoder_nll = -discretized_gaussian_log_likelihood(
+            x_start, means=out["mean"], log_scales=0.5 * out["log_variance"]
+        )
+        if not (decoder_nll.shape == x_start.shape):
+            raise AssertionError
+        decoder_nll = mean_flat(decoder_nll) / np.log(2.0)
+
+        # At the first timestep return the decoder NLL,
+        # otherwise return KL(q(x_{t-1}|x_t,x_0) || p(x_{t-1}|x_t))
+        output = torch.where((t == 0), decoder_nll, kl)
+        return {
+            "output": output,
+            "pred_xstart": out["pred_xstart"],
+            "out_mean": out["mean"],
+            "true_mean": true_mean,
+        }
+
+    def _prior_gaussian(self, x_start: Tensor) -> Tensor:
+        """
+        Get the prior KL term for the variational lower-bound, measured in
+        bits-per-dim.
+
+        This term can't be optimized, as it only depends on the encoder.
+
+        :param x_start: the [N x C x ...] tensor of inputs.
+        :return: a batch of [N] KL values (in bits), one per batch element.
+        """
+        batch_size = x_start.shape[0]
+        t = torch.tensor([self.num_timesteps - 1] * batch_size, device=x_start.device)
+        qt_mean, _, qt_log_variance = self.gaussian_q_mean_variance(x_start, t)
+        kl_prior = normal_kl(
+            mean1=qt_mean, logvar1=qt_log_variance, mean2=0.0, logvar2=0.0
+        )
+        return mean_flat(kl_prior) / np.log(2.0)
+
+    def _gaussian_loss(
+        self,
+        model_out: Tensor,
+        x_start: Tensor,
+        x_t: Tensor,
+        t: Tensor,
+        noise: Tensor,
+        model_kwargs: Any = None,
```

Comment:
I think this would have a problem that if we modify `model_kwargs` inplace, this will change the default value of `model_kwargs`.

Reply:
But for every call that uses `model_kwargs`, you do

``` 
if model_kwargs is None:
    model_kwargs = {}
```
            anyway

Reply:
Yes, but the difference is that if the default `model_kwargs` is `{}`, then when I add an item in this dict, e.g `model_kwargs[k] = v`, the next time I can this method, the default value of `model_kwargs` is no longer `{}` but `{k: v}`. Actually it will create no problem as long as I do not mess around with the default value. This is just a matter of habit to ensure that the default value never changes.